### PR TITLE
UCX/IB/UD: Add UCX_UD_MAX_WINDOW param

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
## What
Introduce UCX_UD_MAX_WINDOW parameter, which makes upper limit of UD congestion avoidance algorithm configurable.

## Why ?
Allow users configure different window size to reduce network congestion.

### CODE REVIEW

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -90,17 +90,18 @@ static void uct_ud_ep_ca_drop(uct_ud_ep_t *ep)
...
-    if (ep->ca.cwnd < UCT_UD_CA_MAX_WINDOW) {
+    if (ep->ca.cwnd < ep->ca.wmax) {
         ep->ca.cwnd += UCT_UD_CA_AI_VALUE;
     }
     ep->tx.max_psn = ep->tx.acked_psn + ep->ca.cwnd;
 }
 
 
-static void uct_ud_ep_reset(uct_ud_ep_t *ep)
+static void uct_ud_ep_reset(uct_ud_ep_t *ep, uct_ud_iface_t *iface)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: can we get iface from ep struct?

File: src/uct/ib/ud/base/ud_iface.c [hunk 2]
```diff
@@ -575,6 +576,11 @@ ucs_config_field_t uct_ud_iface_config_table[] = {
      "Enable checking destination GID for incoming packets of Ethernet network.\n"
      "Mismatched packets are silently dropped.",
      ucs_offsetof(uct_ud_iface_config_t, dgid_check), UCS_CONFIG_TYPE_BOOL},
+
+    {"MAX_WINDOW", UCS_PP_MAKE_STRING(UCT_UD_CA_MAX_WINDOW),
+     "Max congestion avoidance window.",
+     ucs_offsetof(uct_ud_iface_config_t, wmax), UCS_CONFIG_TYPE_UINT},
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: wmax->max_window

File: src/uct/ib/ud/base/ud_iface.h [hunk 3]
```diff
@@ -145,6 +146,7 @@ struct uct_ud_iface {
         unsigned             max_inline;
         int                  check_grh_dgid;
         unsigned             gid_len;
+        unsigned             wmax;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.h]: wmax->max_window

File: src/uct/ib/ud/base/ud_iface.c [hunk 4]
```diff
@@ -439,6 +439,7 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_ud_iface_ops_t *ops, uct_md_h md,
     self->config.peer_timeout    = ucs_time_from_sec(config->peer_timeout);
     self->config.check_grh_dgid  = config->dgid_check &&
                                    uct_ib_iface_is_roce(&self->super);
+    self->config.wmax            = config->wmax;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: check that the value is valid (in range min..max)
of not use ucs_diag() to print error

File: src/uct/ib/ud/base/ud_iface.c [hunk 5]
```diff
@@ -440,6 +440,15 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_ud_iface_ops_t *ops, uct_md_h md,
...
                                    uct_ib_iface_is_roce(&self->super);
 
+    if ((config->max_window < UCT_UD_CA_MIN_WINDOW) ||
+        (config->max_window > UCT_UD_CA_MAX_WINDOW)) {
+        ucs_error("Max congestion avoidance window should be >= %d and <= %d (%d)",
+                  UCT_UD_CA_MIN_WINDOW, UCT_UD_CA_MAX_WINDOW, config->max_window);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    self->config.max_window      = config->max_window;
```
Comment (brminich) [inline src/uct/ib/ud/base/ud_iface.c]: indentation looks strange here

File: src/uct/ib/ud/base/ud_iface.c [hunk 6]
```diff
@@ -575,6 +584,11 @@ ucs_config_field_t uct_ud_iface_config_table[] = {
      "Enable checking destination GID for incoming packets of Ethernet network.\n"
      "Mismatched packets are silently dropped.",
      ucs_offsetof(uct_ud_iface_config_t, dgid_check), UCS_CONFIG_TYPE_BOOL},
+
+    {"MAX_WINDOW", UCS_PP_MAKE_STRING(UCT_UD_CA_MAX_WINDOW),
+     "Max congestion avoidance window.",
```
Comment (brminich) [inline src/uct/ib/ud/base/ud_iface.c]: plz document the valid range for the window, since you print error when it is out of "valid range"

**Comment by shamisp** [comment]
Please add why/what/how

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 05-Mar-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/5355e61c5936d4c0d447f67006972de879405562/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/11eb93c15b65d8a78832b6ff06adf4508625ac0f/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/5e9406e465b2a87a941da5e45c6ef938aafd380c/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/97e8754645e6bf7af11ae6dcd7a32a637c3f32b7/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/3bf9a387a0dc33e4428ea736af9b45a396ad692d/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/77aa9d0c06cc9cc70b5ab3ad3140856c66105587/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/acb23ca86a469f531c0a1e7243b98c056cea8444/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/ae1539837775341664ea9c87561fe71bb494df22/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/28e72e1f924f9f93eda84a9f4773f6dcc929fff2/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/ad36c254ade2d9459e9bf1d44916d2620d42318c/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/0e24e8318a459a71e1c4e4c782f7398be99774cd/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/5d654a74804b43c9b9e265b1b13ec28ccf71df2c/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/48bff64b2db41bb0306595c30b28565359ba998d/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/9ffe365f7e1f46b54fb4087b91c08a7fbf26a707/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/f3aae2fd806318bb2443f9ab1248d216e16d2db1/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/59d160cf21ec3cfa5e0cba93079b7db77041ddf4/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/4272debf884bd6ac855bc11706cb3385d950a621/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/16249a84b9daf0eb750240326634c53c684ba140/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/e075e55fa4849259fbeeccb1e02cfc5b87ec63ce/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/fc0e1b8a0c46bffc28ae4f747ce725057f0342f8/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/32c2648ef1a4bc3578b8aecd9ee42dad95f6df14/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/aeaac9d0171dc2334d2238d27ebf56a8ca12d96e/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/70e10db0b2fdbae62ddf621d7f32c611aecad03a/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/de5f5e880ef3523d05ca885faad409a1ba60c52e/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/3a69063e34b7ab66e9a94070d9924bf8/raw/8409e3f8421cf2e38e3e0e6915c3a8917ec9a2af/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Pull request merged**